consensus/XDPoS/XDPoS.go

```diff
@@ -280,6 +283,9 @@ func (x *XDPoS) GetMasternodesByNumber(chain consensus.ChainReader, blockNumber
 func (x *XDPoS) YourTurn(chain consensus.ChainReader, parent *types.Header, signer common.Address) (int, int, int, bool, error) {
 	switch x.config.BlockConsensusVersion(parent.Number) {
 	case params.ConsensusEngineVersion2:
+		if parent.Number.Cmp(x.config.XDPoSV2Block) == 0 { // TODO: Discuss
+			x.initialV2(chain, parent)
```

Comment:
initialV2 is used two ways: here initialV2(chain, parent).
In `prepare`: initialV2(chain, header).
I suggest to use parent in `prepare` as well. And change how `initialV2` use header. (e.g. change `header.parent` to `header.Hash()`

Reply:
I have another thought, we only need to initialize on Yourturn part, Prepare only happens before Yourturn.